ovs-p4rt/sidecar/testing/prepare_l2_to_tunnel_test.cc

```diff
@@ -0,0 +1,198 @@
+
+static ::p4::config::v1::P4Info p4info;
+
+//----------------------------------------------------------------------
+// PrepareL2ToTunnelTest
+//----------------------------------------------------------------------
+class PrepareL2ToTunnelTest : public ::testing::Test {
+ protected:
+  static void SetUpTestSuite() {
+    // Initialize p4info message from file.
+    ::util::Status status =
+        stratum::ReadProtoFromTextFile(P4INFO_FILE_PATH, &p4info);
+    if (!status.ok()) {
+      std::exit(EXIT_FAILURE);
+    }
+  }
+
+  void DumpMessageAsJson(const google::protobuf::Message& message) {
+    JsonPrintOptions options;
+    options.add_whitespace = true;
+    options.preserve_proto_field_names = true;
+    std::string output;
+    ASSERT_TRUE(MessageToJsonString(message, &output, options).ok());
+    std::cout << output << std::endl;
+  }
+};
+
+uint32_t DecodeWordValue(const std::string string_value) {
+  uint32_t word_value;
+  for (int i = 0; i < string_value.size(); i++) {
+    word_value = (word_value << 8) | string_value[i];
+  }
+  return word_value;
+}
+
+//----------------------------------------------------------------------
+// GetL2ToTunnelV4TableEntry
+//----------------------------------------------------------------------
+constexpr uint8_t kMacAddrV4[6] = {0xd, 0xe, 0xa, 0xd, 0, 0};
+constexpr uint32_t kIpAddrV4 = 0x0a010203U;  // 10.1.2.3
+
+// TODO(derek): How stable are the table and action IDs?
+constexpr uint32_t kTableIdV4 = 43337754U;
+constexpr uint32_t kActionIdV4 = 23805991U;
+
+void get_v4_fdb_info(mac_learning_info& fdb_info) {
+  memcpy(fdb_info.mac_addr, kMacAddrV4, sizeof(fdb_info.mac_addr));
+  fdb_info.tnl_info.remote_ip.family = AF_INET;
+  fdb_info.tnl_info.remote_ip.ip.v4addr.s_addr = kIpAddrV4;
+}
+
+TEST_F(PrepareL2ToTunnelTest, GetL2ToTunnelV4TableEntry) {
+  // Arrange
+  struct mac_learning_info fdb_info = {0};
+  get_v4_fdb_info(fdb_info);
+
+  // Act
+  ::p4::v1::TableEntry table_entry;
+  DiagDetail detail;
+  PrepareL2ToTunnelV4(&table_entry, fdb_info, p4info, true, detail);
+#ifdef DUMP_JSON
+  DumpMessageAsJson(table_entry);
+#endif
+
+  // Assert
+  ASSERT_EQ(table_entry.table_id(), kTableIdV4);
+
+  // match: mac_addr
+  ASSERT_EQ(table_entry.match_size(), 1);
+
+  auto match = table_entry.match()[0];
+  ASSERT_EQ(match.field_id(), 1);
+
+  ASSERT_TRUE(match.has_exact());
+
+  auto match_value = match.exact().value();
+  ASSERT_EQ(match_value.size(), sizeof(kMacAddrV4));
+
+  for (int i = 0; i < sizeof(kMacAddrV4); i++) {
+    ASSERT_EQ(uint8_t(match_value[i]), kMacAddrV4[i]);
+  }
+
+  // action: set_tunnel_v4(dst_addr)
+  ASSERT_TRUE(table_entry.has_action());
+
+  auto table_action = table_entry.action();
+  auto action = table_action.action();
+  ASSERT_EQ(action.action_id(), kActionIdV4);
+
+  ASSERT_EQ(action.params_size(), 1);
+
+  auto param = action.params()[0];
+  ASSERT_EQ(param.param_id(), 1);
+
+  auto param_value = param.value();
+  ASSERT_EQ(param_value.size(), 4);
```

Comment:
I am not very sure about this value.
but looking at sample p4info.txt for LNWv3, i can see IPv4 has one param only

actions {
  preamble {
    id: 23805991
    name: "linux_networking_control.set_tunnel_v4"
    alias: "set_tunnel_v4"
  }
  **params {**
    id: 1
    name: "dst_addr"
    bitwidth: 32
  }
}

But same IPv6 has 4 params.
actions {
  preamble {
    id: 23953453
    name: "linux_networking_control.set_tunnel_v6"
    alias: "set_tunnel_v6"
  }
  params {
    id: 1
    name: "ipv6_1"
    bitwidth: 32
  }
  params {
    id: 2
    name: "ipv6_2"
    bitwidth: 32
  }
  params {
    id: 3
    name: "ipv6_3"
    bitwidth: 32
  }
  params {
    id: 4
    name: "ipv6_4"
    bitwidth: 32
  }
}